### DISCUSSION TITLE
Mobile apps (Official and Toot!) unable to sign-in/search for instances in LIMITED_FEDERATION_MODE

### DISCUSSION THREAD

**Discussion by Fireshaded** [mastodon/mastodon#17383] [Q&A]
Not sure if this is a design choice or something no one has really come across since instances running in LIMITED_FEDERATION_MODE are rare.  But if someone tries to add a server they're running that is in LIMITED_FEDERATION_MODE, they will be unable to find the server by entering the URL.  If the instance is put to normal mode, user logs in, then put back into LIMITED_FEDERATION_MODE, they're able to use the app normally.

I'm guessing this is since the API and other endpoints are locked down when in LIMITED_FEDERATION_MODE.

Anyway to open up the API or make adjustments to the Official Mastodon app to allow connecting to instances that are in LIMITED_FEDERATION_MODE instead of having to turn on and off the function for each user doing the initial connection?

Also can anyone confirm this is happening and not some weird problem with the instances I run?

Thanks!

**Comment by pfrybar** [comment]
Yes! I ran into this same issue today and have been banging my head against the wall until I finally figured out it's due to `LIMITED_FEDERATION_MODE`. I am looking to run a private instance for my family (I know, I know, it goes against the model) and the killer feature which would get them to use it would be mobile app(s).

It seems like this wouldn't be too hard to "fix".

Edit: oops, this isn't an answer, but rather a confirmation of your issue.

Edit 2: @Fireshaded based on your discovery that things work if you first put the app into "normal" mode, then sign in, and finally turn `LIMITED_FEDERATION_MODE` back on, I was able to come up with a workaround for my own private server. What I've done is to stub out the `/api/v1/instance` endpoint which is used to discover the private instance and return a static json response which is public. My nginx conf and static json is listed below. It would be great if this could be changed/fixed in the codebase itself.

nginx conf:
```
    location /api/v1/instance {
        alias /path/to/instance.json;
    }
```

json response (instance.json):
```
{
  "uri": "<domain>",
  "title": "None",
  "short_description": "None",
  "description": "None",
  "email": "none@example.com",
  "version": "3.0.0",
  "urls": {
    "streaming_api": "wss://<domain>"
  },
  "stats": {
    "user_count": 0,
    "status_count": 0,
    "domain_count": 0
  },
  "thumbnail": "",
  "languages": [
    "en"
  ],
  "registrations": false,
  "approval_required": true,
  "contact_account": {
    "id": "1",
    "username": "None",
    "acct": "None",
    "display_name": "None",
    "locked": false,
    "bot": false,
    "created_at": "2000-01-01T00:00:00.000Z",
    "note": "None",
    "url": "https://example.com",
    "avatar": "https://example.com",
    "avatar_static": "https://example.com",
    "header": "https://example.com",
    "header_static": "https://example.com",
    "followers_count": 0,
    "following_count": 0,
    "statuses_count": 0,
    "last_status_at": "2000-01-01T00:00:00.000Z",
    "emojis": [],
    "fields": []
  }
}
```

**Comment by Fireshaded** [comment]
Oh wow! Thanks for the information and the "fix".  It would be nice if there is an option or just possible to have the apps connect to the instance when in Limited_Federation_Mode, especially with internal instances or ones used by families.

Having the app is great and it becomes a pain to have to either set times (thus exposing the instance to the public) to turn off Limited_Federation_Mode so people can join or do it one by one. Kinda ruins the ease and ability to get people set up without hickups when trying to get people onboard with open source projects.

**Comment by Westley-Winks** [comment]
For anyone in the future, I fixed this by exposing `/api/v1/instance` (and `/api/v2/instance` by inheritance) to the public even when in `LIMITED_FEDERATION_MODE`. See [this commit](https://github.com/Westley-Winks/heyjohn/commit/47bcb59210f79a87f085c15fbd8f7249f7a56039) to see what lines I modified.

It does the same thing as @pfrybar 's solution but does it from the codebase instead. Some apps call those endpoints to get the description and make sure it is actually a Mastodon server. Good luck!
